File: elixir_implementation/jux/lib/jux/parser.ex
defmodule Jux.Parser do
  @moduledoc """
  Parses a string into a list of Jux tokens that can then be applied to a stack.
  """

  @whitespace_regexp ~r{\A\s+}m
  @comment_regexp ~r{^#.*}
  @float_regexp ~r{^[+-]?\d+\.\d+}
  @integer_regexp ~r{^[+-]?\d+}
  @identifier_regexp ~r{^[a-zA-Z_][\w.]*[?!]?}

  def valid_identifier?(str) do
    str =~ @identifier_regexp
  end

  @doc """
  Takes a string as input, returns a list of Jux tokens as output.
  Notice that the output of the parser is not yet reversed. 
  Before evaluation, this should happen; the tail of the list of tokens
  should be evaluated before the head.
  """
  def parse(str) do
    do_parse(str, [])
  end

  defp do_parse("", tokens), do: tokens
  defp do_parse(source, tokens) do
    #IO.inspect {source, tokens}
    cond do
      source =~ @whitespace_regexp ->
        [_, rest] = remove_token_from_string(source, @whitespace_regexp)
        do_parse(rest, tokens)
      source =~ @comment_regexp ->
        [_, rest] = remove_token_from_string(source, @comment_regexp)
        do_parse(rest, tokens)
      String.starts_with?(source, "[") ->
        [quotation, rest] = parse_quotation(source)
        do_parse(rest, [quotation | tokens])
      String.starts_with?(source, "\"") ->
        [string, rest] = parse_string(source)
        do_parse(rest, [string | tokens])
      source =~ @identifier_regexp ->
        [identifier, rest] = remove_token_from_string(source, @identifier_regexp)
        do_parse(rest, [parse_identifier(identifier) | tokens])
      source =~ @float_regexp ->
        [float, rest] = remove_token_from_string(source, @float_regexp)
        do_parse(rest, [parse_float(float) | tokens])
      source =~ @integer_regexp ->
        [integer, rest] = remove_token_from_string(source, @integer_regexp)
        do_parse(rest, [parse_integer(integer) | tokens])
      :otherwise ->
        raise "Could not parse rest of source: #{inspect source}"
    end
  end

  defp remove_token_from_string(source, regex) do 
    [prefix | _] = Regex.run(regex, source)
    #IO.inspect(prefix)
    #IO.inspect(String.replace_prefix(source, prefix, ""))
    [prefix, String.replace_prefix(source, prefix, "")]
  end


  def parse_quotation(source) do
    [quotation_length, rest] = do_parse_quotation(String.next_codepoint(source), 0, 0)
    quotation = String.slice(source, 1, quotation_length - 2)
    [parse(quotation), rest]
  end

  defp do_parse_quotation(nil, _, _) do
    raise "unmatched `[` in input."
  end

  defp do_parse_quotation({"[", rest}, bracket_count, length_acc) do
    # IO.puts "Position: #{length_acc}"
    do_parse_quotation(String.next_codepoint(rest), bracket_count+1, length_acc+1)
  end

  defp do_parse_quotation({"]", rest}, 1, length_acc) do
    [length_acc+1, rest]
  end

  defp do_parse_quotation({"]", rest}, bracket_count, length_acc) do
    do_parse_quotation(String.next_codepoint(rest), bracket_count-1, length_acc+1)
  end

  defp do_parse_quotation({_, rest}, bracket_count, length_acc) do
    do_parse_quotation(String.next_codepoint(rest), bracket_count, length_acc+1)
  end


  def parse_string(source) do
    [string_length, rest] = do_parse_string(String.next_codepoint(source), 0)
    string = String.slice(source, 1, string_length - 2)
    [string, rest]
  end

  defp do_parse_string({"\"", rest}, 0) do
    do_parse_string(String.next_codepoint(rest), 1)
  end

  defp do_parse_string({"\"", rest}, length) do
    [length+1, rest]
  end

  defp do_parse_string({"\\", rest}, length) do
    case String.next_codepoint(rest) do
      {"\"", restrest} ->
        do_parse_string(String.next_codepoint(restrest), length+2)
      {chars, _} ->
        do_parse_string(String.next_codepoint(rest), length+1)
    end
  end

  defp do_parse_string({_, rest}, length) do
    do_parse_string(String.next_codepoint(rest), length+1)
  end


  def parse_identifier(str) do
    # atom = str |> String.to_existing_atom
    # res = Jux.Stdlib.__info__(:functions)[atom]
    # if res != 1 do
    #   raise ArgumentError, "unknown identifier `#{str}`"
    # end
    # atom
    Jux.Identifier.new(str)
    #apply(Jux.Stdlib, atom, [stack])
  end

  def parse_integer(str) do
      str
      |> String.to_integer
  end

  def parse_float(str) do
      str
      |> String.to_float
  end

end

File: ruby_implementation/jux/parser.rb
module Jux
  class Parser
    class << self
      def whitespace_regexp
        %r{\A\s+}m
      end

      def comment_regexp
        %r{\A#.*}
      end

      def float_regexp
        %r{\A[+-]?\d+\.\d+}
      end

      def integer_regexp
        %r{\A[+-]?\d+}
      end

      def identifier_regexp
        %r{\A[a-zA-Z_][\w.]*[?!]?}
      end

      def escaped_identifier_regexp
        %r{\A/[a-zA-Z_][\w.]*[?!]?}
      end

      def parse(str)
        token_str = str
        function_queue = []

        loop do
          case
          when whitespace_str = token_str.match(whitespace_regexp)
            token_str = token_str[whitespace_str.length..-1]
          when comment_str = token_str.match(comment_regexp)
            token_str = token_str[comment_str.to_s.length..-1]
          when token_str.match(/^\[/)
            quotation, token_str = parse_quotation(token_str)
            function_queue << Jux::Token.new(quotation, "Quotation")
          when token_str.match(/^"/)
            str, token_str = parse_string(token_str)
            function_queue << str
          when escaped_identifier_str = token_str.match(escaped_identifier_regexp)
            function_queue << Jux::Token.new(Jux::EscapedIdentifier.new(escaped_identifier_str.to_s[1..-1]), "Identifier")
          	puts "EscapedIdentifier #{escaped_identifier_str}; Left token str: #{token_str.inspect}"
            token_str = token_str[escaped_identifier_str.to_s.length..-1]
          when identifier_str = token_str.match(identifier_regexp)
            function_queue << Jux::Token.new(Jux::Identifier.new(identifier_str.to_s), "Identifier")
          	puts "Identifier #{identifier_str}; Left token str: #{token_str.inspect}"
            token_str = token_str[identifier_str.to_s.length..-1]
          when integer_str = token_str.match(integer_regexp)
            integer = integer_str[0].to_i
            function_queue << Jux::Token.new(integer, "Integer")
            token_str = token_str[integer_str.to_s.length..-1]
          when token_str == ""
            break
          else
            raise "Improper Jux syntax: `#{token_str}`"
          end
        end
        puts Jux::Helper.stack_to_str(function_queue)  
        function_queue
      end

      def parse_string(token_str)
        i = 1
        loop do
          raise "unmatched \"" if token_str[i].nil?
          break if token_str[i] == '"'
          i+= 1 if token_str[i..(i+1)] == '\"'
          i+= 1
        end
        new_str = Jux::Helper.ruby_str_to_jux_str(token_str[1...i])
        token_str = token_str[(i+1)..-1]
        [new_str, token_str]
      end

      def parse_quotation(token_str)
        i = 0
        depth = 0
        loop do
          raise "unmatched [" if token_str[i].nil?
          if token_str[i] == "["
            depth += 1
          elsif token_str[i] == "]"
            break if depth == 1
            depth -= 1
          end
          i += 1
        end

        quotation = parse(token_str[1...i])
        token_str = token_str[(i+1)..-1]
        [quotation, token_str]
      end

      def valid_identifier?(str)
        str =~ %r{^[a-zA-Z_][\w.]*[?!]?$}
      end
    end
  end
end
